File: leetcode/python/000125.valid-palindrome.py
class Solution:
	# Time complexity: O(n)
	# Space complexity: O(1)
	def isPalindrome(self, s: str) -> bool:
		s = s.lower()
		l, r = 0, len(s) - 1
		while l <= r:
			if not s[l].isalnum():
				l += 1
				continue
			if not s[r].isalnum():
				r -= 1
				continue
			if s[l] != s[r]:
				return False
			l += 1
			r -= 1
		return True

File: leetcode/javascript/000125.valid-palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 * Time complexity: O(n)
 * Space complexity: O(1)
 */
var isPalindrome = function (s) {
  let l = 0;
  let r = s.length - 1;

  while (l <= r) {
    if (!/[a-zA-Z0-9]/.test(s[l])) {
      l++;
      continue;
    }
    if (!/[a-zA-Z0-9]/.test(s[r])) {
      r--;
      continue;
    }
    if (s[l].toLowerCase() !== s[r].toLowerCase()) {
      return false;
    }
    l++;
    r--;
  }

  return true;
};
